File: WordIndexer/Java/src/WordIndexer_v1_1.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;
import java.util.Map.Entry;
import java.util.SortedMap;
import java.util.TreeMap;

public class WordIndexer_v1_1 {
	public static void main(String[] args) {
		String text =	"The cat cat sat sat on the mat\n" + "The dog chased the cat\n" + "The cat ran from the dog";
		String[] omittedWords = {"the","on","from"}; 
		
		for(Iterator<Entry<String, List<Integer>>> index = generateIndex(text, omittedWords).entrySet().iterator(); index.hasNext();) {
			Entry<String, List<Integer>> entry = index.next();
			System.out.println(entry.getKey() + " " + entry.getValue());
		}
	}
	
	public static SortedMap<String, List<Integer>> generateIndex(String text, String[] omittedWordsArray) {
		List<String> omittedWords = new ArrayList<String>(Arrays.asList(omittedWordsArray));
		SortedMap<String, List<Integer>> indexMap = new TreeMap<String, List<Integer>>();		
		List<String> lines = Arrays.asList(text.split("\n"));
		
		int lineNumber = 1;
		for(Iterator<String> line = lines.iterator(); line.hasNext(); lineNumber++) {
			List<String> words = Arrays.asList(line.next().toLowerCase().replaceAll("[^A-Za-z0-9 ]+", "").split("\\s+"));
			for(Iterator<String> word = words.iterator(); word.hasNext();) {
				String key = word.next();
				if(!omittedWords.contains(key)) {
					List<Integer> tempSet = new ArrayList<Integer>();
					if(indexMap.containsKey(key)) {
						tempSet = indexMap.get(key);
					}
					tempSet.add(lineNumber);
					indexMap.put(key, tempSet);
				}
			}
		}
		return indexMap;
	}
	
	public static SortedMap<String, List<Integer>> generateIndex(String text) {
		String[] omittedWords = new String[0];
		return generateIndex(text, omittedWords);
	}
}

File: WordIndexer/Python/src/WordIndexer_v1_1.py
from collections import defaultdict, OrderedDict
import re

def generate_index(text, omitted_words=[]):
    index = defaultdict(list)
    for i, line in enumerate(re.sub("[^\w\n\s]+", "", text.lower()).split("\n")):
        for word in line.split():
            if word not in omitted_words and (i+1) not in index[word]:
                index[word].append(i+1)
    return OrderedDict(sorted(index.items()))

text =  "The cat cat sat on the mat\n" + "The dog chased the cat\n" + "The cat ran from the dog"
omitted_words = ["the","on","from"]

index = generate_index(text, omitted_words)
for word, appearances in index.items(): 
    print(word, appearances)